### PR TITLE
fix(pty): make failed websocket attachments reapable

### LINKED ISSUE #110849: PtySession.attach() commits attached state before the replay write — a client dropping during attach leaves the PTY permanently unreapable
### Summary

`PtySession.attach()` commits the attached state *before* it writes the replay buffer to the socket, and the caller (`pty_ws`) performs that call **outside** any `try/finally`. A client that drops during the attach replay therefore raises out of the handler with the session already marked `attached=True, last_detached_at=None` and never reaches `PTY_REGISTRY.detach()`.

That state is the one state `reap_idle()` can never reclaim:

* `not s.alive` is `False` — the PTY child is perfectly alive;
* the second clause requires `not s.attached` **and** `s.last_detached_at is not None` — both are false.

`_reap_one_idle_or_raise()` has the same `not attached` requirement, and the 30-minute TTL is anchored to `last_detached_at`, which stays `None` forever. The PTY process group (`node ui-tui/dist/entry.js` + its `python -m tui_gateway.entry` child) then lives indefinitely, holding a tty, the PTY master fd, its loopback sockets and the session's work.

This is an independent entry into "attached forever" — it needs no reverse proxy, no disabled WS ping, and no SIGHUP cascade; the client only has to disappear between `attach` bookkeeping and the first byte of the replay.

### Where (verified on `main` @ a89c1e11)

`hermes_cli/pty_session.py` (attach, ~L98-L114):

```python
if self._ws is not ws:
    await _close_ws(self._ws, WS_CLOSE_SUPERSEDED)
self._ws = ws
self._attach_generation += 1
self.attached = True
self.last_detached_at = None
if snap := self.buffer.snapshot():
    await ws.send_bytes(snap)          # <- raises here, state already committed
```

`hermes_cli/web_routers/chat_ws.py` (L516) — the call site is outside the `try` that owns the detach; `PTY_REGISTRY.detach()` only runs in the writer loop's `finally` further down:

```python
if not await session.attach(ws, force_redraw=not _created):
    await _close_stalled_pty_input(ws, path="keepalive-redraw")
    PTY_REGISTRY.detach(attach_token, ws)
    return
```

### Evidence

Dashboard server log (3 occurrences, same shape), client-side `/api/pty` socket:

```
chat_ws.py:516 in pty_ws
pty_session.py:111 in attach
    await ws.send_bytes(snap)
starlette/websockets.py:89 in send -> raise WebSocketDisconnect(code=1006)
...
uvicorn.protocols.utils.ClientDisconnected
```

Field accumulation on one `hermes dashboard` with a single real terminal (`who` → console + one SSH session):

* peak snapshot: **18** `node ui-tui/dist/entry.js` and **15** `python -m tui_gateway.entry` processes, **15** ttys held, **15** PTY master fds in the dashboard process;
* all ESTABLISHED sockets on the dashboard port were held by the dashboard process itself (no browser clients at all) — i.e. no live viewer anywhere, yet every session stayed attached;
* one orphaned `ui-tui` node (no `tui_gateway` child) survived from the same day five days earlier and was still present in every later snapshot;
* a rescan 45 minutes later showed the total fluctuating (8 nodes / 7 children) but the oldest orphan unchanged — the leak is cumulative, not a one-off.

### Impact

Unbounded process / PTY / fd leak per dropped client, plus indirectly a stuck session lease: a stranded `tui_gateway` with an unfinished turn keeps refreshing its `session_turn_leases` row, which is invisible to watchdog checks keyed on "lease expired" or "owner pid dead" (the expiry is fresh and the pid is alive).

### Proposed fix (any of)

1. In `attach()`, do not commit `_ws`/`attached`/`last_detached_at` until the replay buffer has been delivered — send first, then record the attachment (or roll back in `except`).
2. In `pty_ws`, wrap the `session.attach(...)` call in `try/except` and `PTY_REGISTRY.detach(attach_token, ws)` before re-raising (or fold the attach into the writer loop's `try/finally`).
3. In `reap_idle()`, add a criterion independent of `attached` (e.g. an attach timestamp / socket-liveness probe) so a session whose viewer is gone is eventually reclaimed even if the detach was never observed.

Regression test: a `FakeWS` whose `send_bytes` raises → after the failed attach the session must still be reclaimable (`attached is False` or `last_detached_at is not None`). The existing suite pins the opposite invariant (`tests/test_pty_session.py` — "attached → not reapable") but does not cover an aborted attach.

### Related

* #110164 — orphaned `ui-tui` nodes with dead gateway children; the attach-abort path above is another, proxy-independent source of the same accumulation.
* #96418 — attached session never reaped when the loopback bind disables the WS ping (external trigger); this report is the in-process trigger and does not require the ping to be off.
* #106400 — `reap_idle()` `pop(key)` without default; fixed on `main` (a89c1e11) and **not** the same path as this report.
* PR #110168 — keeps the embedded dashboard TUI alive on SIGHUP; unrelated to the stale-attached state.

### PR DESCRIPTION
**Pull request by fangliquanflq**
## Summary

- detach the current websocket generation when buffered replay fails
- detach on live drain send failures while preserving a superseding viewer
- cover both failure paths and their replacement-socket races

## Why this PR vs existing PR #110859 and #110861

PR #110861 handles only replay failure and explicitly leaves the live drain failure path unresolved. PR #110859 handles both send sites, but cleanup is guarded only by socket identity and its tests do not exercise failures that finish after a replacement attaches. This PR scopes cleanup to the captured attachment generation, covers task cancellation during replay through `finally`, and tests that stale replay and drain failures cannot detach the replacement viewer.

## Test plan

- `scripts/run_tests.sh tests/hermes_cli/test_pty_session.py`

## Related Issue

Closes #110849

### CODE REVIEW

**Comment by alt-glitch** [comment]
> *This was generated by AI during triage.*

Competing fix for #110849 alongside #110859 and #110861. This PR scopes detach to the captured attachment generation and covers both replay and live-drain failures plus replacement-socket races; #110859 covers both sites guarded by socket identity; #110861 covers replay only. Flagging the three-way cluster so a maintainer picks one.

**Comment by fangliquanflq** [comment]
Thanks for flagging the competing-fix cluster. I compared the current diffs: #110861 handles replay failure only, while #110859 handles replay and live-drain failures with socket-identity cleanup but does not exercise failures completing after a replacement socket attaches. This PR already scopes cleanup to both the captured socket and attachment generation, covers replay cancellation via `finally`, and tests replacement-socket races on both failure paths, so this is informational and no code change is needed.

Verification: `scripts/run_tests.sh tests/hermes_cli/test_pty_session.py` — 16 passed.

**Comment by kyssta-exe** [comment]
## Review: generation-scoped detach for failed attachments

**Summary:** The most complete of the three #110849 fixes (vs #110859/#110861): both failure paths (attach-replay and live `_drain`) detach through a generation-scoped `_detach_if_current`, and the replay path is wrapped in try/finally so task cancellation mid-replay also detaches. Tests cover the replacement-socket races the other PRs don't.

**What changed**
- `hermes_cli/pty_session.py`: `_attach_generation` counter; `attach()` captures its generation and detaches in `finally` unless completed; `_drain` failures detach only if socket+generation still current; `detach()` delegates to `_detach_if_current`.
- `tests/hermes_cli/test_pty_session.py`: stale-replay and stale-drain race tests proving a replacement viewer survives.

**Strengths**
- Generation scoping (not just socket identity) is the correct guard: it also covers the blocked-send race where the old socket object is still referenced but superseded.
- `finally`-based cleanup covers cancellation, which bare excepts miss — the subtle, correct choice for async websocket code.
- The race tests (blocked failing socket + replacement attach + release) are exactly the scenario that would regress a naive fix.

**Findings**
- Verify `_attach_generation` is initialized in `__init__` (the visible hunks only show it incremented/read) and that `Optional` is imported for the `_detach_if_current` signature — both likely fine, just not visible in the reviewed hunks.
- **Coordination needed:** overlaps #110859/#110861 on the same lines; the three cannot all merge. Of the three, this one subsumes the others — suggest landing this and closing the rest.

**Verdict:** Looks good to merge (my pick of the three), after confirming init/import and the dedupe decision.

*Reviewed using Hermes-Agent*
